flow/device_flow.go

```diff
@@ -0,0 +1,200 @@
+
+	// RequestedAudience contains the access token audience as requested by the OAuth 2.0 Client.
+	//
+	// required: true
+	RequestedAudience sqlxx.StringSliceJSONFormat `db:"requested_at_audience"`
+
+	// Client is the OAuth 2.0 Client that initiated the request.
+	//
+	// required: true
+	Client *client.Client `db:"-"`
+
+	ClientID string `db:"client_id"`
+	Verifier string `db:"verifier"`
+	CSRF     string `db:"csrf"`
+
+	RequestedAt time.Time `db:"requested_at"`
+
+	State int16 `db:"state"`
+	// The user_code was already handled.
+	UserCodeWasUsed bool `db:"user_code_was_used"`
+	// HandledAt contains the timestamp the device user verification request was handled.
+	UserCodeHandledAt sqlxx.NullTime      `db:"user_code_handled_at"`
+	Error             *RequestDeniedError `db:"error"`
+}
+
+func NewDeviceFlow(r *DeviceUserAuthRequest) *DeviceFlow {
+	return &DeviceFlow{
+		ID:                r.ID,
+		Client:            r.Client,
+		ClientID:          r.ClientID,
+		RequestURL:        r.RequestURL,
+		Verifier:          r.Verifier,
+		CSRF:              r.CSRF,
+		RequestedAt:       r.RequestedAt,
+		UserCodeWasUsed:   r.WasHandled,
+		UserCodeHandledAt: r.HandledAt,
+		State:             DeviceFlowStateInitialized,
+	}
+}
+
+func (f *DeviceFlow) GetDeviceUserAuthRequest() *DeviceUserAuthRequest {
+	return &DeviceUserAuthRequest{
+		ID:          f.ID,
+		Client:      f.Client,
+		ClientID:    f.ClientID,
+		RequestURL:  f.RequestURL,
+		Verifier:    f.Verifier,
+		CSRF:        f.CSRF,
+		RequestedAt: f.RequestedAt,
+		WasHandled:  f.UserCodeWasUsed,
+		HandledAt:   f.UserCodeHandledAt,
+	}
+}
+
+func (f *DeviceFlow) GetHandledDeviceUserAuthRequest() *HandledDeviceUserAuthRequest {
+	return &HandledDeviceUserAuthRequest{
+		ID:                f.ID,
+		Client:            f.Client,
+		ClientID:          f.ClientID,
+		Request:           f.GetDeviceUserAuthRequest(),
+		RequestURL:        f.RequestURL,
+		RequestedAt:       f.RequestedAt,
+		RequestedScope:    f.RequestedScope,
+		RequestedAudience: f.RequestedAudience,
+		WasHandled:        f.UserCodeWasUsed,
+		HandledAt:         f.UserCodeHandledAt,
+		Error:             f.Error,
+	}
+}
+
+func (f *DeviceFlow) HandleDeviceUserAuthRequest(h *HandledDeviceUserAuthRequest) error {
+	if f.UserCodeWasUsed {
+		return errors.WithStack(x.ErrConflict.WithHint("The user_code was already used and can no longer be changed."))
+	}
+
+	if f.State != DeviceFlowStateInitialized && f.State != DeviceFlowStateUnused && f.State != DeviceFlowStateError {
+		return errors.Errorf("invalid flow state: expected %d/%d/%d, got %d", DeviceFlowStateInitialized, DeviceFlowStateUnused, DeviceFlowStateError, f.State)
+	}
+
+	if f.ID != h.ID {
+		return errors.Errorf("flow device challenge ID %s does not match HandledDeviceUserAuthRequest ID %s", f.ID, h.ID)
+	}
+
+	if h.Error != nil {
+		f.State = DeviceFlowStateError
+	} else {
+		f.State = DeviceFlowStateUnused
+	}
```

Comment:
```suggestion
	f.State = DeviceFlowStateUnused
	
	if h.Error != nil {
		f.State = DeviceFlowStateError
	} 	
	
	
```

maybe put https://github.com/canonical/hydra/pull/3/files#diff-7949c4faf556a5c2f0e80176c7ab205478dfff7edc4f5b794215201632c36f05R146 at the end of the func, like line 156